### PR TITLE
Add fix for evo2 generate/inference

### PR DESCRIPTION
**Pull request by jwilber**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Fixes Evo 2 inference issue where state wasn't correctly accumulating during generate. 

Background: 
Documented issue example: https://github.com/NVIDIA/bionemo-framework/issues/890

**Collection**: [Note which collection this PR will affect]
GPT/hyena
# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/gpt/model/megatron/hyena/hyena_mixer.py [hunk 1]
```diff
@@ -280,18 +291,38 @@ def forward(self, x, layer_past=None, inference_context=None, _hyena_use_cp=True
...
+                pad_len = (multiple - (seq_len % multiple)) % multiple
+                if pad_len == 0:
+                    return x
+                from torch.nn.functional import pad
+                return pad(x, (0, 0, 0, pad_len))  # Pad sequence length dimension
+            L = x.shape[0]
+            x = rearrange(x, "l b d -> b l d")
+            x = pad_to_multiple(x)
+            x = rearrange(x, "b l d -> l b d")
+            features, _ = self._maybe_use_fp8(self.dense_projection, x)
```
Comment (farhadrgh) [inline nemo/collections/llm/gpt/model/megatron/hyena/hyena_mixer.py]: Would this work? to avoid using multiple rearrangements

```suggestion
            def pad_to_multiple(x, multiple=16):
                """Pad tensor to make sequence length divisible by multiple."""
                seq_len = x.size(0)
                if seq_len % multiple == 0:
                    return x
                
                pad_len = multiple - (seq_len % multiple)
                pad_tensor = torch.zeros(pad_len, *x.shape[1:], device=x.device, dtype=x.dtype)
                return torch.cat([x, pad_tensor], dim=0)
            
            # Direct padding without rearrange
            x = pad_to_multiple(x)
            features, _ = self._maybe_use_fp8(self.dense_projection, x)
```

File: nemo/collections/llm/gpt/model/megatron/hyena/engine.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
+# Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (farhadrgh) [inline nemo/collections/llm/gpt/model/megatron/hyena/engine.py:1]: The target code coverage is 80% in NeMo. You might need some tests to hit the target for this file

File: nemo/collections/llm/gpt/model/megatron/hyena/hyena_mixer.py (hunk 1 above)
Comment (jwilber) [reply nemo/collections/llm/gpt/model/megatron/hyena/hyena_mixer.py]: Good call

File: nemo/collections/llm/gpt/model/megatron/hyena/hyena_mixer.py [hunk 3]
```diff
@@ -160,7 +162,15 @@ def __init__(
 
         hyena_proj_groups = self.proj_groups if not self.grouped_attention else 1
         grouped_proj_size = self.hidden_size_per_partition // hyena_proj_groups
-        self.hyena_proj_conv = ParallelCausalDepthwiseConv1d(
+        hyena_proj_conv_class = ParallelCausalDepthwiseConv1d
+        if with_state_p:
```
Comment (jstjohn) [inline nemo/collections/llm/gpt/model/megatron/hyena/hyena_mixer.py]: IMO let's get rid of this since it's hard-coded to True. That was prob an implementation hack for verifying things.

File: nemo/collections/llm/gpt/model/megatron/hyena/hyena_mixer.py [hunk 4]
```diff
@@ -160,7 +162,15 @@ def __init__(
...
+            hyena_proj_conv_class = ParallelCausalDepthwiseConv1dWithState
+
+        short_conv_class = ParallelCausalDepthwiseConv1d
+        if with_state_s:
```
Comment (jstjohn) [inline nemo/collections/llm/gpt/model/megatron/hyena/hyena_mixer.py]: Same here, let's get rid of this.

File: nemo/collections/llm/gpt/model/megatron/hyena/hyena_mixer.py [hunk 5]
```diff
@@ -36,12 +36,14 @@
 from nemo.collections.llm.gpt.model.megatron.hyena.hyena_utils import (
     B2BCausalConv1dModule,
     ParallelCausalDepthwiseConv1d,
+    ParallelCausalDepthwiseConv1dWithState,
     ParallelHyenaOperator,
     ParallelShortHyenaOperator,
     divide,
+    with_state_p,
+    with_state_s,
```
Comment (jstjohn) [inline nemo/collections/llm/gpt/model/megatron/hyena/hyena_mixer.py]: And these, since they're always True.

File: nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py [hunk 6]
```diff
@@ -39,6 +39,11 @@
 
 from nemo.collections.llm.gpt.model.megatron.hyena.hyena_config import HyenaConfig
 
+with_state_p = True
+with_state_s = True
+with_state_m = True
+with_state_l = True
+
```
Comment (jstjohn) [inline nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py]: Let's get rid of these and just use the new classes like we are now since they're all hard-coded to True.

File: nemo/collections/llm/gpt/model/megatron/hyena/hyena_mixer.py [hunk 7]
```diff
@@ -160,7 +160,13 @@ def __init__(
 
         hyena_proj_groups = self.proj_groups if not self.grouped_attention else 1
         grouped_proj_size = self.hidden_size_per_partition // hyena_proj_groups
-        self.hyena_proj_conv = ParallelCausalDepthwiseConv1d(
+
+        hyena_proj_conv_class = ParallelCausalDepthwiseConv1dWithState
+
+
```
Comment (JRD971000) [inline nemo/collections/llm/gpt/model/megatron/hyena/hyena_mixer.py]: Same class getting instantiated

File: nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py [hunk 8]
```diff
@@ -862,6 +971,13 @@ def forward(self, x1, x2, v, _hyena_use_cp=True):
 
         h = h.repeat_interleave(self.group_dim, dim=-2)
 
+        if self.operator_type == "hyena_medium_conv":
```
Comment (JRD971000) [inline nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py]: add condition on inference

File: nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py [hunk 9]
```diff
@@ -829,7 +941,7 @@ def forward(self, x1, x2, v, _hyena_use_cp=True):
         # The kernel length must be adjusted in CP settings
         _L_kernel = L if cp_group is None else L * len(torch.distributed.get_process_group_ranks(cp_group))
         if self.use_medium_hyena:
-            h = self.filter(min(self.hyena_medium_conv_len, _L_kernel))
+            h = self.filter(self.hyena_medium_conv_len)
```
Comment (jstjohn) [inline nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py:944]: Was this intentional? Does it work without this?

File: nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py [hunk 10]
```diff
@@ -862,23 +974,29 @@ def forward(self, x1, x2, v, _hyena_use_cp=True):
...
+            z = x1 * z
 
-        if cp_group is not None and len(torch.distributed.get_process_group_ranks(cp_group)) > 1:
-            z = AllToAllSingleFunction.apply(z, cp_group, "full_to_split", True)
-            # [ B, H, L // num_ranks]
-        return z  # [B, (G, DG), L]
+            if cp_group is not None and len(torch.distributed.get_process_group_ranks(cp_group)) > 1:
+                z = AllToAllSingleFunction.apply(z, cp_group, "full_to_split", True)
+                # [ B, H, L // num_ranks]
+            return z  # [B, (G, DG), L]
```
Comment (jstjohn) [inline nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py:999]: This is fine as is, but you _could_ move this code block into a third forward called something like `forward_train(...)` and return that in this case, then the intention is a bit more self documenting.

**Review by jstjohn — DISMISSED** [review]
Approved with some cosmetic comments and one code block diff to verify.

File: nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py (hunk 9 above)
Comment (jwilber) [reply nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py:944]: Yes, without this the results all drop to ~10%!

File: nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py (hunk 10 above)
Comment (jstjohn) [reply nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py:999]: Feel free to mark this resolved with whatever you decide.

**Review by JRD971000 — APPROVED** [review]
LGTM, we discussed it over a call with @jstjohn. Thanks Jared!

**Review by JRD971000 — DISMISSED** [review]

**Comment by chtruong814** [comment]
Although the code coverage is below the targeted patch change, we will go ahead and merge this given the need to handle a support request. This has a minimal impact for the coverage of the llm collection overall.
